jumanji/environments/logic/sudoku/reward.py

```diff
@@ -0,0 +1,39 @@
+# Copyright 2022 InstaDeep Ltd. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import abc
+
+import chex
+import jax.numpy as jnp
+
+from jumanji.environments.logic.sudoku.types import State
+from jumanji.environments.logic.sudoku.utils import puzzle_completed
+
+
+class RewardFn(abc.ABC):
```

Comment:
I think that's just the contributing guidelines here, @clement-bonnet ?

Reply:
It's to make it like a protocol (which you could alternatively use instead). So that users may define their own reward functions as long as they inherit from `RewardFn`.

Reply:
But do they have to inherit from `RewardFn` ? 
If we had `RewardFn = Callable[[State, chex.Array, State, jnp.ndarray], chex.Array]` you have the same protocol, but the user has the freedom to either pass a function or an instance of a class implementing `__call__`. Especially in this case where the reward is super simple, it could just be `reward_fn = lambda *_, new_state, d:  puzzle_completed(new_state.board).astype(float)`

But I just noticed it is also defined for other env, so better keeping it like that for consistency :+1:

Reply:
It is true it might look overkill for this use case, happy to keep it this way for consistency though.